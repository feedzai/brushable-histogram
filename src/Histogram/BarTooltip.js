import React from "react";
import PropTypes from "prop-types";
import Portal from "../common/components/Portal";
import { HISTOGRAM_BAR_TOOLTIP_WIDTH, HISTOGRAM_BAR_TOOLTIP_HEIGHT } from "../constants";

/**
 * BarTootip
 *
 * Renders the histogram bar tooltip that allows to display information about a specific bar.
 * The tooltip content is customizable and is provided by the tooltipBarCustomization custom render function.
 *
 * @author Nuno Neves (nuno.neves@feedzai.com)
 */

const propsTypes = {
    selectedBarPosition: PropTypes.shape({
        top: PropTypes.number,
        left: PropTypes.number,
        width: PropTypes.number
    }),
    currentBar: PropTypes.object.isRequired,
    tooltipBarCustomization: PropTypes.func.isRequired
};
const defaultProps = {
    selectedBarPosition: {
        top: 0,
        left: 0,
        width: 0
    }
};

function BarTooltip({ selectedBarPosition, currentBar, tooltipBarCustomization }) {
    if (typeof tooltipBarCustomization !== "function") {
        return null;
    }

    const { top, left, width } = selectedBarPosition;
    const tooltipStyle = {
        position: "fixed",
        left: `${left - (HISTOGRAM_BAR_TOOLTIP_WIDTH / 2) + (width / 2)}px`,
        top: `${top - HISTOGRAM_BAR_TOOLTIP_HEIGHT}px`
    };

    const tooltipElement = tooltipBarCustomization(currentBar);

    return (
        <Portal>
            <div
                className="fdz-css-graph-histogram-bars--tooltip"
                style={tooltipStyle}
            >
                {tooltipElement}
            </div>
        </Portal>
    );
}

BarTooltip.propTypes = propsTypes;
BarTooltip.defaultProps = defaultProps;

export default React.memo(BarTooltip);
