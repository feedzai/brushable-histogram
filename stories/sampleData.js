const NUMBER_OF_POINTS = 60000;
const startTimestamp = 1533164400000;

/**
 * Calculates some sample data
 *
 * @param {number} numberOfPoints
 * @returns {Array.<Object>}
 */
export function calculate(numberOfPoints) {
    const data = [];

    for (let i = 0; i < numberOfPoints; i++) {
        data.push({
            "timestamp": startTimestamp + Math.abs(Math.floor(Math.cos(i) * 1000000000)),
            "total": Math.abs(Math.sin(i))
        });
    }

    return data;
}

export default calculate(NUMBER_OF_POINTS);

export const smallSample = calculate(10);
