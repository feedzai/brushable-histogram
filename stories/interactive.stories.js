/**
 * 2019, Feedzai
 *
 * This source code is licensed under the LGPL license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from "react";
import { storiesOf } from "@storybook/react";
import { withKnobs, object } from "@storybook/addon-knobs";
import Histogram from "../src/index";
import "../src/Histogram/Histogram.scss";

/**
 * @author Victor Fernandes (victor.fernandes@feedzai.com)
 * @since 1.7.0
 */

const stories = storiesOf("Histogram - Interactive data", module);

stories.addDecorator(withKnobs);

stories
    .add("Basic example", () => (
        <Histogram
            data={object("Data", [
                { timestamp: 1170070000000, total: 100 },
                { timestamp: 1270070000000, total: 23 },
                { timestamp: 1370070000000, total: 100 },
                { timestamp: 1470070000000, total: 23 },
                { timestamp: 1570070000000, total: 100 },
                { timestamp: 1670070000000, total: 23 },
                { timestamp: 1770070000000, total: 400 },
                { timestamp: 1870070000000, total: 200 }
            ])}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
            onIntervalChange={(domain) => console.log(domain)}
        />
    ));
