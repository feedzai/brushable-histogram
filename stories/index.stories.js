import React, { Fragment } from "react";
import { timeFormat } from "d3-time-format";

import { storiesOf } from "@storybook/react";
import { withKnobs, number, object } from "@storybook/addon-knobs";

import sampleData from "./sampleData";
import Histogram from "../src/index";
import "../src/Histogram/Histogram.scss";

const stories = storiesOf("Histogram", module);

stories.addDecorator(withKnobs);

const formatDayFull = timeFormat("%e %b'%y");

function formatContextInterval(dateStart, dateEnd) {
    const formatStart = formatDayFull(dateStart);
    const formatEnd = formatDayFull(dateEnd);

    // If the formatted start and end date are the same then just display one
    if (formatStart === formatEnd) {
        return formatStart;
    }

    // Otherwise display the interval limits.
    return `${formatStart} - ${formatEnd}`;
}

function histogramTooltipBar(bar) {
    return (
        <Fragment>
            <div className="fdz-css-graph-histogram-bars--tooltip-value">
                {Math.floor(bar.yValue)} Events
            </div>
            <div className="fdz-css-graph-histogram-bars--tooltip-dates">
                {formatContextInterval(bar.x0, bar.x1)}
            </div>
        </Fragment>
    );
}

stories
    .add("Basic example", () => (
        <Histogram
            data={sampleData}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
        />
    ))
    .add("With a custom tooltip", () => (
        <Histogram
            data={sampleData}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
            tooltipBarCustomization={histogramTooltipBar}
        />
    ))
    .add("With a custom axis formatters", () => (
        <Histogram
            data={sampleData}
            xAccessor={(datapoint) => datapoint.timestamp}
            xAxisFormatter={formatDayFull}
            yAxisFormatter={(value) => (value > 0 ? `${value}$` : "")}
            yAccessor={(datapoint) => datapoint.total}
        />
    ))
    .add("Without the play button", () => (
        <Histogram
            data={sampleData}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
            renderPlayButton={false}
        />
    ))
    .add("With a custom a height", () => (
        <Histogram
            data={sampleData}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
            height={number("height (min 150)", 300)}
            tooltipBarCustomization={histogramTooltipBar}
        />
    ))
    .add("With no data", () => (
        <Histogram
            data={[]}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
            tooltipBarCustomization={histogramTooltipBar}
        />
    ))
    .add("Change data in order to update histogram", () => (
        <Histogram
            data={object("Data", [
                { timestamp: 1170070000000, total: 100 },
                { timestamp: 1270070000000, total: 23 },
                { timestamp: 1370070000000, total: 100 },
                { timestamp: 1470070000000, total: 23 },
                { timestamp: 1570070000000, total: 100 },
                { timestamp: 1670070000000, total: 23 },
                { timestamp: 1770070000000, total: 400 },
                { timestamp: 1870070000000, total: 200 }
            ])}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
            onIntervalChange={(domain) => console.log(domain)}
        />
    ));
