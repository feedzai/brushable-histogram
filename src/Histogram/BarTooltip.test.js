import React, { Fragment } from "react";
import { mount } from "enzyme";
import { timeFormat } from "d3-time-format";
import BarTooltip from "./BarTooltip";
import { smallSample } from "../../stories/sampleData";

/**
 * BarTootip
 *
 * Renders the histogram bar tooltip that allows to display information about a specific bar.
 * The tooltip content is customizable and is provided by the tooltipBarCustomization custom render function.
 *
 * @author Nuno Neves (nuno.neves@feedzai.com)
 */

const formatMinute = timeFormat("%I:%M");

function histogramTooltipBar(bar) {
    return (
        <Fragment>
            <div>
                {bar.yValue} Events
            </div>
            <div>
                {`${formatMinute(bar.x0)} - ${formatMinute(bar.x1)}`}
            </div>
        </Fragment>
    );
}

const selectedBarPosition = {
    top: 100,
    left: 50,
    width: 20
};

describe("Histogram/BarTooltip", () => {
    describe("render", () => {
        test("should return null if props.tooltipBarCustomization is not a function", () => {
            const wrapper = mount(
                <BarTooltip
                    tooltipBarCustomization={null}
                    currentBar={smallSample}
                    selectedBarPosition={selectedBarPosition}
                />
            );

            expect(wrapper).toMatchSnapshot();
        });

        test("should render the tooltip", () => {
            const wrapper = mount(
                <BarTooltip
                    tooltipBarCustomization={histogramTooltipBar}
                    currentBar={smallSample}
                    selectedBarPosition={selectedBarPosition}
                />
            );

            expect(wrapper).toMatchSnapshot();
        });
    });
});
