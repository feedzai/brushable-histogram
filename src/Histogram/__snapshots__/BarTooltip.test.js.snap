// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Histogram/BarTooltip render should render the tooltip 1`] = `
<BarTooltip
  currentBar={
    Array [
      Object {
        "timestamp": 1534164400000,
        "total": 0,
      },
      Object {
        "timestamp": 1533704702305,
        "total": 0.8414709848078965,
      },
      Object {
        "timestamp": 1533580546837,
        "total": 0.9092974268256817,
      },
      Object {
        "timestamp": 1534154392497,
        "total": 0.1411200080598672,
      },
      Object {
        "timestamp": 1533818043621,
        "total": 0.7568024953079282,
      },
      Object {
        "timestamp": 1533448062185,
        "total": 0.9589242746631385,
      },
      Object {
        "timestamp": 1534124570286,
        "total": 0.27941549819892586,
      },
      Object {
        "timestamp": 1533918302254,
        "total": 0.6569865987187891,
      },
      Object {
        "timestamp": 1533309900034,
        "total": 0.9893582466233818,
      },
      Object {
        "timestamp": 1534075530262,
        "total": 0.4121184852417566,
      },
    ]
  }
  selectedBarPosition={
    Object {
      "left": 50,
      "top": 100,
      "width": 20,
    }
  }
  tooltipBarCustomization={[Function]}
>
  <Portal>
    <Portal
      containerInfo={
        <div>
          <div
            class="fdz-css-graph-histogram-bars--tooltip"
            style="position: fixed; left: 10px; top: 60px;"
          >
            <div>
               Events
            </div>
            <div>
              12:NaN - 12:NaN
            </div>
          </div>
        </div>
      }
    >
      <div
        className="fdz-css-graph-histogram-bars--tooltip"
        style={
          Object {
            "left": "10px",
            "position": "fixed",
            "top": "60px",
          }
        }
      >
        <div>
           Events
        </div>
        <div>
          12:NaN - 12:NaN
        </div>
      </div>
    </Portal>
  </Portal>
</BarTooltip>
`;

exports[`Histogram/BarTooltip render should return null if props.tooltipBarCustomization is not a function 1`] = `
<BarTooltip
  currentBar={
    Array [
      Object {
        "timestamp": 1534164400000,
        "total": 0,
      },
      Object {
        "timestamp": 1533704702305,
        "total": 0.8414709848078965,
      },
      Object {
        "timestamp": 1533580546837,
        "total": 0.9092974268256817,
      },
      Object {
        "timestamp": 1534154392497,
        "total": 0.1411200080598672,
      },
      Object {
        "timestamp": 1533818043621,
        "total": 0.7568024953079282,
      },
      Object {
        "timestamp": 1533448062185,
        "total": 0.9589242746631385,
      },
      Object {
        "timestamp": 1534124570286,
        "total": 0.27941549819892586,
      },
      Object {
        "timestamp": 1533918302254,
        "total": 0.6569865987187891,
      },
      Object {
        "timestamp": 1533309900034,
        "total": 0.9893582466233818,
      },
      Object {
        "timestamp": 1534075530262,
        "total": 0.4121184852417566,
      },
    ]
  }
  selectedBarPosition={
    Object {
      "left": 50,
      "top": 100,
      "width": 20,
    }
  }
  tooltipBarCustomization={null}
/>
`;
